### PR TITLE
Xuanzhao Gao, hw10, compressed sensing

### PR DESCRIPTION
**Pull request by ArrogantGao**

### CODE REVIEW

File: courseworks/week10/xuanzhaogao/hw10.ipynb [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+   "metadata": {},
+   "source": [
+    "$$\n",
+    "\\begin{split}\n",
+    "    \\frac{d \\mathcal{L}}{d x} \n",
+    "    & = \\frac{d \\mathcal{L}}{d ||Ax - b||_2} \\frac{d \\sqrt{(Ax - b)^T (Ax - b)}}{d (Ax - b)} \\frac{d (Ax - b)}{dx} \\\\\n",
+    "    & = 1 \\times \\frac{1}{2 ||Ax - b||_2} \\times 2(Ax - b)^T \\times A\\\\\n",
+    "    & = \\frac{(Ax - b)^T \\times A}{||Ax - b||_2}\n",
+    "\\end{split}\n",
+    "$$"
```
Comment (GiggleLiu) [inline courseworks/week10/xuanzhaogao/hw10.ipynb:34]: 1. please use $\partial$ for partial derivative.
2. the shape of the output vector is not correct.
```julia
julia> randn(3)' * randn(3, 3)
1×3 adjoint(::Vector{Float64}) with eltype Float64:
 -1.85937  0.156183  -1.7463
```

The result should be transposed.

File: courseworks/week10/xuanzhaogao/hw10.ipynb [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    "total_length = length(text)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "frequency_count = Dict()\n",
```
Comment (GiggleLiu) [inline courseworks/week10/xuanzhaogao/hw10.ipynb:98]: Please use `Dict{Char, Int}()` to improve type stability.

File: courseworks/week10/xuanzhaogao/hw10.ipynb [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "symbols = \"\"\n",
+    "probs = []\n",
```
Comment (GiggleLiu) [inline courseworks/week10/xuanzhaogao/hw10.ipynb:115]: Please use `Float64[]` to improve type stability.

**Review by GiggleLiu — APPROVED** [review]
After fixing some minor issues, this PR can be merged to the master. Well done.

File: courseworks/week10/xuanzhaogao/hw10.ipynb (hunk 1 above)
Comment (ArrogantGao) [reply courseworks/week10/xuanzhaogao/hw10.ipynb:34]: 1. The notation for the partial derivative have been revised.
2. I think the output vector given as in this way may make more sense, considering the simple relation:
```math
\delta f = \frac{\partial  f}{\partial x} \delta x~,
```
if $\delta f$ is a scalar and  $\delta x$ is a column, then $\frac{\partial  f}{\partial x}$ should be a row.

File: courseworks/week10/xuanzhaogao/hw10.ipynb (hunk 2 above)
Comment (ArrogantGao) [reply courseworks/week10/xuanzhaogao/hw10.ipynb:98]: The types have been specified.

**Pull request merged**